client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,314 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utilities for tracing block execution
+
+use std::{collections::HashMap, sync::{Arc, atomic::{AtomicU64, Ordering}}};
+use std::time::Instant;
+
+use parking_lot::Mutex;
+use tracing::{Dispatch, dispatcher, Subscriber, Level, span::{Attributes, Record, Id}};
+use tracing_subscriber::CurrentSpan;
+
+use sc_client_api::BlockBackend;
+use sc_rpc_server::MAX_PAYLOAD;
+use sp_api::{Core, Metadata, ProvideRuntimeApi, Encode};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use sp_rpc::tracing::{BlockTrace, Span, TraceError, TraceBlockResponse};
+use sp_tracing::{WASM_NAME_KEY, WASM_TARGET_KEY, WASM_TRACE_IDENTIFIER};
+use sp_core::hexdisplay::HexDisplay;
+use crate::{SpanDatum, TraceEvent, Values};
+
+// Estimate of the max size base RPC payload when the Id is bound as a u64. If strings
+// are used for the RPC Id this may need to be adjusted. Note: The base payload does not
+// include the RPC result.
+const BASE_RPC_PAYLOAD: usize = 500;
+// Default to only pallet, frame support and state related traces
+const DEFAULT_TARGETS: &'static str = "pallet,frame,state";
+const TRACE_TARGET: &'static str = "block_trace";
+// Default to not filtering based on storage keys as storage keys vary per chain.
+const DEFAULT_STORAGE_KEYS: &'static str = "";
+// The name of a field required for all events.
+const REQUIRED_EVENT_FIELD: &'static str  = "method";
+
+struct BlockSubscriber {
+	targets: Vec<(String, Level)>,
+	next_id: AtomicU64,
+	current_span: CurrentSpan,
+	spans: Mutex<HashMap<Id, SpanDatum>>,
+	events: Mutex<Vec<TraceEvent>>,
+}
+
+impl BlockSubscriber {
+	fn new(
+		targets: &str,
+		spans: Mutex<HashMap<Id, SpanDatum>>,
+		events: Mutex<Vec<TraceEvent>>,
+	) -> Self {
+		let next_id = AtomicU64::new(1);
+		let mut targets: Vec<_> = targets
+			.split(',')
+			.map(|s| crate::parse_target(s))
+			.collect();
+		// Ensure that WASM traces are always enabled
+		// Filtering happens when decoding the actual target / level
+		targets.push((WASM_TRACE_IDENTIFIER.to_owned(), Level::TRACE));
+		BlockSubscriber {
+			targets,
+			next_id,
+			current_span: CurrentSpan::default(),
+			spans,
+			events,
+		}
+	}
+}
+
+impl Subscriber for BlockSubscriber {
+	fn enabled(&self, metadata: &tracing::Metadata<'_>) -> bool {
+		for (target, level) in &self.targets {
+			if metadata.target().starts_with(target.as_str()) && metadata.level() <= level {
+				if metadata.is_span() {
+					return true;
+				} else if metadata.fields().field(REQUIRED_EVENT_FIELD).is_some() {
+					// it is an event, so we check if has the method field
+					return true;
+				}
+			}
+		}
+		false
+	}
+
+	fn new_span(&self, attrs: &Attributes<'_>) -> Id {
+		let id = self.next_id.fetch_add(1, Ordering::SeqCst);
```

Comment:
I think this could be `Ordering::Relaxed` but it's doesn't matter that much because it's only faster on weakly-ordered platforms such as ARM

Reply:
We actually want the Ids to correspond to the order the spans are entered. 
My concern with using anything but the strictest ordering is that this assumption could break, although I have a poor understanding of async memory access so I could be wrong.

Reply:
The id still atomic unless your using the ID for other synchronization such as indexing into memory or something similar.

But if you are not sure then you did the right call to make is `SeqCst` :)

Reply:
I don't see any reason here as well to not use Relaxed.

Atomic ordering just instructs the compiler to insert barriers, that certain operations are done in a particular order. However, as here no ordering is required, relaxed should be enough.

And this also should be faster on x86_64, because they also support parallel execution of op codes.

Reply:
> I don't see any reason here as well to not use Relaxed.

@bkchr Would there be an issue if this function was called by a few separate threads at the same time? From what you are saying it sounds like not matter what atomic ordering is chosen it would not have an affect

Reply:
You need certain ordering of operations for stuff where something relies on the fact that a certain variable is set before you lock/unlock. However in your case here you don't require such ordering. You only lock to insert into the hashmap and at this point the operation that has incremented the Atomic variable was already executed because it is required here.

Reply:
Makes sense, thanks for the explanation :)